comms/src/protocol/messaging/mod.rs

```diff
@@ -0,0 +1,275 @@
+                        self.handle_request(req).await,
+                        "Failed to handle request because '{error}'",
+                    );
+                },
+                notification = self.proto_notification.select_next_some() => {
+                    self.handle_notification(notification).await;
+                },
+                complete => {
+                    info!(target: LOG_TARGET, "MessagingProtocol is shutting down because all streams have completed");
+                    break;
+                }
+            }
+        }
+    }
+
+    pub fn framed<TSubstream>(socket: TSubstream) -> Framed<IoCompat<TSubstream>, LengthDelimitedCodec>
+    where TSubstream: AsyncRead + AsyncWrite + Unpin {
+        Framed::new(IoCompat::new(socket), LengthDelimitedCodec::new())
+    }
+
+    async fn handle_request(&mut self, req: MessagingRequest) -> Result<(), MessagingProtocolError> {
+        use MessagingRequest::*;
+        match req {
+            SendMessage(node_id, msg, reply_tx) => {
+                let _ = reply_tx.send(self.send_message(node_id, msg).await);
+            },
+        }
+
+        Ok(())
+    }
+
+    async fn send_message(&mut self, node_id: Box<NodeId>, msg: Bytes) -> Result<(), MessagingProtocolError> {
+        let sender = match self.active_queues.entry(node_id.clone()) {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => {
+                match self
+                    .connection_manager_requester
+                    .get_active_connection(*node_id.clone())
+                    .await
+                    .map_err(MessagingProtocolError::UnexpectedConnectionManagerError)?
+                {
+                    Some(conn) => {
+                        let sender = Self::spawn_outbound_handler(self.executor.clone(), conn).await?;
+                        entry.insert(sender)
+                    },
+                    None => {
+                        // Should have an active connection before sending a message
+                        error!(
+                            target: LOG_TARGET,
+                            "Attempted to send a message to peer '{}' that is not connected", node_id
+                        );
+                        return Err(MessagingProtocolError::PeerNotConnected);
+                    },
+                }
+            },
+        };
+
+        let mut attempts = 0;
+        loop {
+            match sender.send(msg.clone()).await {
+                Ok(_) => {
+                    return Ok(());
+                },
+                Err(err) => {
+                    // Lazily remove Senders from the active queue if the MessagingProtocolHandler has shut down
+                    if err.is_disconnected() {
+                        self.active_queues.remove(&node_id);
+                        break;
+                    }
+
+                    attempts += 1;
+                    if attempts > 5 {
```

Comment:
This threshold should probably come from a const or a config

Reply:
True - I did kind of bang that retry functionality in there (after trying a [way to do it properly](https://gist.github.com/sdbondi/cbc43752faa75914b36ae9dd3fd13915)) - I think that this case should resolve pretty quickly or not happen at all so this is just a quick way to improve reliability on the chance that it happens. So I vote for a const in the file vs a configuration field - will update the code